csharp/extractor/Semmle.Extraction.CIL/Entities/Type.cs

```diff
@@ -92,35 +75,44 @@ public override IEnumerable<IExtractionProduct> Contents
             }
         }
 
-        /// <summary>
-        /// Whether this type is visible outside the current assembly.
-        /// </summary>
-        public virtual bool IsVisible => true;
-
         public abstract string Name { get; }
 
-        public abstract Namespace? Namespace { get; }
+        public abstract Namespace? ContainingNamespace { get; }
 
         public abstract Type? ContainingType { get; }
 
+        public virtual TypeContainer Parent => (TypeContainer?)ContainingType ?? ContainingNamespace!;
+
         public abstract Type Construct(IEnumerable<Type> typeArguments);
 
         /// <summary>
-        /// The number of type arguments, or 0 if this isn't generic.
-        /// The containing type may also have type arguments.
+        /// Returns the type arguments of constructed types. For non-constructed types it returns an
+        /// empty collection.
+        /// </summary>
+        public virtual IEnumerable<Type> ThisTypeArguments
+        {
+            get
+            {
+                yield break;
+            }
+        }
+
+        /// <summary>
+        /// The number of type parameters for non-constructed generic types, the number of type arguments
+        /// for constructed types, or 0.
         /// </summary>
-        public abstract int ThisTypeParameters { get; }
+        public abstract int ThisTypeParameterCount { get; }
 
         /// <summary>
-        /// The total number of type parameters (including parent types).
+        /// The total number of type parameters/type arguments (including parent types).
         /// This is used for internal consistency checking only.
         /// </summary>
-        public int TotalTypeParametersCheck =>
-            ContainingType == null ? ThisTypeParameters : ThisTypeParameters + ContainingType.TotalTypeParametersCheck;
+        public int TotalTypeParametersCount => ContainingType == null
+            ? ThisTypeParameterCount
+            : ThisTypeParameterCount + ContainingType.TotalTypeParametersCount;
```

Comment:
```suggestion
        public int TotalTypeParametersCount =>
            ThisTypeParameterCount + (ContainingType?.TotalTypeParametersCount ?? 0);
```
A bit shorter, don't know what you prefer.